core-primitives/sgx/crypto/src/rsa3072.rs

```diff
@@ -64,56 +65,88 @@ impl ShieldingCryptoEncrypt for Rsa3072PubKey {
 	}
 }
 
+impl ToPubkey for Rsa3072KeyPair {
+	type Error = Error;
+	type Pubkey = Rsa3072PubKey;
+
+	fn pubkey(&self) -> Result<Self::Pubkey> {
+		self.export_pubkey().map_err(|e| Error::Other(format!("{:?}", e).into()))
+	}
+}
+
+pub trait RsaSealing {
+	fn unseal_pubkey(&self) -> Result<Rsa3072PubKey>;
+
+	fn unseal_pair(&self) -> Result<Rsa3072KeyPair>;
+
+	fn exists(&self) -> bool;
+
+	fn create_sealed_if_absent(&self) -> Result<()>;
+
+	fn create_sealed(&self) -> Result<()>;
+}
+
 #[cfg(feature = "sgx")]
 pub mod sgx {
 	use super::*;
-	use derive_more::Display;
+	use crate::key_repository::KeyRepository;
 	use itp_settings::files::RSA3072_SEALED_KEY_FILE;
-	use itp_sgx_io::{seal, unseal, SealedIO, StaticSealedIO};
+	use itp_sgx_io::{seal, unseal, SealedIO};
 	use log::*;
-	use std::sgxfs::SgxFile;
+	use std::{path::PathBuf, sgxfs::SgxFile};
+
+	/// Gets a repository for an Rsa3072 keypair and initializes
+	/// a fresh key pair if it doesn't exist at `path`.
+	pub fn get_rsa3072_repository(
+		path: PathBuf,
+	) -> Result<KeyRepository<Rsa3072KeyPair, Rsa3072Seal>> {
+		let rsa_seal = Rsa3072Seal::new(path);
+		rsa_seal.create_sealed_if_absent()?;
+		let shielding_key = rsa_seal.unseal_pair()?;
+		Ok(KeyRepository::new(shielding_key, rsa_seal.into()))
+	}
+
+	#[derive(Clone, Debug)]
+	pub struct Rsa3072Seal {
+		base_path: PathBuf,
+	}
 
 	impl Rsa3072Seal {
-		pub fn unseal_pubkey() -> Result<Rsa3072PubKey> {
-			let pair = Self::unseal_from_static_file()?;
-			let pubkey =
-				pair.export_pubkey().map_err(|e| Error::Other(format!("{:?}", e).into()))?;
-			Ok(pubkey)
+		pub fn new(base_path: PathBuf) -> Self {
+			Self { base_path }
 		}
-	}
 
-	pub fn create_sealed_if_absent() -> Result<()> {
-		if SgxFile::open(RSA3072_SEALED_KEY_FILE).is_err() {
-			info!("[Enclave] Keyfile not found, creating new! {}", RSA3072_SEALED_KEY_FILE);
-			return create_sealed()
+		pub fn path(&self) -> PathBuf {
+			self.base_path.join(RSA3072_SEALED_KEY_FILE)
 		}
-		Ok(())
 	}
 
-	pub fn create_sealed() -> Result<()> {
-		let rsa_keypair =
-			Rsa3072KeyPair::new().map_err(|e| Error::Other(format!("{:?}", e).into()))?;
-		// println!("[Enclave] generated RSA3072 key pair. Cleartext: {}", rsa_key_json);
-		Rsa3072Seal::seal_to_static_file(&rsa_keypair)
-	}
+	impl RsaSealing for Rsa3072Seal {
+		fn unseal_pubkey(&self) -> Result<Rsa3072PubKey> {
+			self.unseal()?.pubkey()
+		}
 
-	#[derive(Copy, Clone, Debug, Display)]
-	pub struct Rsa3072Seal;
+		fn unseal_pair(&self) -> Result<Rsa3072KeyPair> {
+			self.unseal()
+		}
 
-	impl StaticSealedIO for Rsa3072Seal {
-		type Error = Error;
-		type Unsealed = Rsa3072KeyPair;
-		fn unseal_from_static_file() -> Result<Self::Unsealed> {
-			let raw = unseal(RSA3072_SEALED_KEY_FILE)?;
-			let key: Rsa3072KeyPair = serde_json::from_slice(&raw)
-				.map_err(|e| Error::Other(format!("{:?}", e).into()))?;
-			Ok(key.into())
+		fn exists(&self) -> bool {
```

Comment:
Can we not just call exists on the `PathBuf` directly?

`self.path().exists()`

Reply:
Hmm, yes. The semantics are slightly different; if the file exists, and we can't open it, your version returns true, while mine doesn't, but I think this is ok and actually more precise considering the naming of the function. I shall change it.